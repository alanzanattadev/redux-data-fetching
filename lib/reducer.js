"use babel";
// @flow

import { fromJS, Map, List, Set, Record } from "immutable";
import { normalize, schema } from "normalizr";
import {
  graphQLizr,
  convertsNormalizedEntitiesToRecords,
  convertsGraphQLResultToRootEntitiesIDs,
} from "./graphqlTypesConverters";
import type { GraphQLSchema as GraphQLSchemaType } from "graphql";

export const ResultsRecord = Record({ byQuery: Map(), byEntity: Map() });
export const QueryRecord = Record({ results: new ResultsRecord() });
export const DataReducerRecord = Record({
  entities: Map(),
  queries: Map(),
});

export default function configureReducer(
  normalizrTypes: { [typeName: string]: schema.Entity },
  recordsTypes: { [typeName: string]: Record<*> },
  graphQLSchema: GraphQLSchemaType,
) {
  function warnBadIDRequest(type: string, supposedId: mixed): void {
    console.warn(
      "You're trying to delete a key of bad type for type",
      type,
      ":",
      supposedId,
      ".Keys must be of type number, string or identified object eg: {id: 'key'}. Aborting request",
    );
  }

  function getID(data: string | number | { id: string }): ?string {
    if (typeof data === "string") return data;
    else if (typeof data === "number") return data.toString();
    else if (typeof data === "object" && data !== null) {
      return getID(data.id);
    } else {
      return null;
    }
  }

  return function reducer(
    state: DataReducerRecord = new DataReducerRecord(),
    action: { type: string, payload: Object },
  ) {
    switch (action.type) {
      case "DATA_RECEIVED":
        const normalizrModel = Object.keys(
          action.payload.entities,
        ).reduce((red, key) => {
          const type = normalizrTypes[key];
          if (type === undefined)
            throw new Error(
              "You can't normalize a type which is not an Entity. An Entity is a type with an id attribut. You may have defined a GraphQL root query type with a route that has a type without any id.",
            );
          return Object.assign({}, red, {
            [key]:
              typeof action.payload.entities[key] == "object" &&
              Array.isArray(action.payload.entities[key])
                ? [type]
                : type,
          });
        }, {});
        const normalized = normalize(
          JSON.parse(JSON.stringify(action.payload.entities)),
          normalizrModel,
        );
        return state
          .update("entities", entities =>
            entities.mergeDeepWith(
              (a, b) => (b === undefined ? a : b),
              convertsNormalizedEntitiesToRecords(
                normalized.entities,
                recordsTypes,
                graphQLSchema,
              ),
            ),
          )
          .update(
            "queries",
            queries =>
              action.payload.query
                ? queries.set(
                    action.payload.query.request.hash,
                    new QueryRecord({
                      results: new ResultsRecord({
                        byQuery: convertsGraphQLResultToRootEntitiesIDs(
                          action.payload.query.response.raw,
                        ),
                      }),
                    }),
                  )
                : queries,
          );
      case "DATA_REMOVED":
        return Object.keys(action.payload).reduce((red, key: string) => {
          if (
            typeof action.payload[key] == "object" &&
            Array.isArray(action.payload[key])
          ) {
            return action.payload[key].reduce((reduction, value) => {
              const id = getID(value);
              if (id == null) {
                warnBadIDRequest(key, value);
                return reduction;
              } else {
                return reduction.deleteIn(["entities", key, id]);
              }
            }, red);
          } else {
            const id = getID(action.payload[key]);
            if (id == null) {
              warnBadIDRequest(key, action.payload[key]);
              return red;
            } else {
              // $FlowFixMe
              return red.deleteIn(["entities", key, id]);
            }
          }
        }, state);
      default:
        return state;
    }
  };
}
