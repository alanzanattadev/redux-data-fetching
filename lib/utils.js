// @flow

import { Stack, List } from "immutable";
import {
  TypeInfo,
  GraphQLSchema,
  Kind,
  GraphQLList,
  GraphQLObjectType,
  getNamedType,
  visit,
  visitWithTypeInfo,
  parse,
} from "graphql";
import type { ASTNode, FieldNode } from "graphql";
import { DataReducerRecord } from "./reducer";
import { isEntity } from "./graphqlTypesConverters";

export function hashString(str: string): number {
  let hash = 0;
  if (str.length == 0) return hash;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash = hash & hash; // Convert to 32bit integer
  }
  return hash;
}

export class TypeInfoWithValuesComparator extends TypeInfo {
  _value1Stack: Stack<mixed>;
  _value2Stack: Stack<mixed>;
  _reducer1: DataReducerRecord;
  _reducer2: DataReducerRecord;
  _queryHash1: string | number;
  _queryHash2: string | number;
  _shouldBreak: boolean;

  constructor({
    schema,
    reducer1,
    reducer2,
    queryHash1,
    queryHash2,
  }: {
    schema: GraphQLSchema,
    reducer1: DataReducerRecord,
    reducer2: DataReducerRecord,
    queryHash1: string | number,
    queryHash2: string | number,
  }) {
    super(schema);
    this._value1Stack = Stack();
    this._value2Stack = Stack();
    this._reducer1 = reducer1;
    this._reducer2 = reducer2;
    this._queryHash1 = queryHash1;
    this._queryHash2 = queryHash2;
    this._shouldBreak = false;
  }

  _getQueryValue(
    node: FieldNode,
    reducer: DataReducerRecord,
    queryHash: string | number,
  ) {
    if (reducer == null || queryHash == null)
      throw new Error(
        "Trying to determine whether selection of data has to be relaunched, but encounterd an error. Missing dependency on reducer or queryHash",
      );
    const queryResult = reducer.getIn([
      "queries",
      queryHash,
      "results",
      "byQuery",
      node.alias != null ? node.alias.value : node.name.value,
    ]);
    const currentType = this.getType();
    if (queryResult == null) {
      if (currentType instanceof GraphQLList) return [];
      else return null;
    }
    if (currentType instanceof GraphQLList) {
      const typeName = getNamedType(currentType.ofType).name;
      return queryResult.map(id =>
        reducer.getIn(["entities", typeName, id.toString()]),
      );
    } else if (queryResult != null) {
      const namedType = getNamedType(currentType);
      if (namedType != undefined) {
        return reducer.getIn([
          "entities",
          namedType.name,
          queryResult.toString(),
        ]);
      } else {
        console.error("CurrentType:", currentType);
        throw new Error(
          "Weird behavior has been encountered with an unnamed type as current type",
        );
      }
    } else {
      return null;
    }
  }

  _getEntityValue(
    node: FieldNode,
    reducer: DataReducerRecord,
    lastValue: mixed,
  ) {
    if (
      node.selectionSet != null &&
      typeof lastValue === "object" &&
      lastValue != null
    ) {
      const fieldName = node.alias ? node.alias.value : node.name.value;
      if (Array.isArray(lastValue) || List.isList(lastValue))
        // $FlowFixMe
        return lastValue.map(v => this._getEntityValue(node, reducer, v));
      const fieldValue = lastValue[fieldName];
      const namedType = getNamedType(this.getType());
      if (namedType && isEntity(namedType) === false) {
        return fieldValue;
      }
      if (
        this.getType() instanceof GraphQLList &&
        fieldValue != null &&
        (Array.isArray(fieldValue) || List.isList(fieldValue))
      ) {
        if (namedType) {
          const entityName = namedType.name;
          // $FlowFixMe
          return fieldValue.map(id =>
            reducer.getIn(["entities", entityName, id.toString()]),
          );
        } else {
          console.error(
            "Weird state encountered trying to get the named type of",
            this.getType(),
            "at node",
            node,
            "knowing that ",
            fieldValue,
            "is an Array or a Immutable.List",
          );
          throw new Error("Aborting.");
        }
      } else if (fieldValue != null) {
        if (namedType) {
          const entityName = namedType.name;
          // $FlowFixMe
          return reducer.getIn(["entities", entityName, fieldValue.toString()]);
        } else {
          console.error(
            "Weird state encountered trying to get the named type of",
            this.getType(),
            "at node",
            node,
            "knowing that ",
            fieldValue,
            "is an Object or a Immutable.Record",
          );
          throw new Error("Aborting.");
        }
      } else {
        return null;
      }
    } else {
      return lastValue;
    }
  }

  _haveChanged(value1: mixed, value2: mixed): boolean {
    if (typeof value1 !== typeof value2) return true;
    if (Array.isArray(value1) || List.isList(value1)) {
      if (Array.isArray(value2) === true || List.isList(value2) === true) {
        // $FlowFixMe
        return value1.some((v1: mixed, i: number) => {
          return this._haveChanged(
            v1,
            // $FlowFixMe
            Array.isArray(value2) ? value2[i] : value2.get(i),
          );
        });
      } else return true;
    } else if (typeof value1 === "object" && value1 !== null) {
      if (value2 === null) return true;
      return value1 !== value2;
    } else {
      return value1 !== value2;
    }
  }

  enter(node: ASTNode) {
    super.enter(node);
    if (node.kind === Kind.FIELD) {
      let value1;
      let value2;
      if (this.getParentType() === this._schema.getQueryType()) {
        value1 = this._getQueryValue(node, this._reducer1, this._queryHash1);
        this._value1Stack = this._value1Stack.unshift(value1);
        value2 = this._getQueryValue(node, this._reducer2, this._queryHash2);
        this._value2Stack = this._value2Stack.push(value2);
      } else {
        value1 = this._getEntityValue(
          node,
          this._reducer1,
          this._value1Stack.peek(),
        );
        this._value1Stack = this._value1Stack.push(value1);
        value2 = this._getEntityValue(
          node,
          this._reducer2,
          this._value2Stack.peek(),
        );
        this._value2Stack = this._value2Stack.push(value2);
      }
      if (this._shouldBreak === false) {
        this._shouldBreak = this._haveChanged(value1, value2);
      }
    }
  }

  leave(node: ASTNode) {
    if (node.kind === Kind.FIELD) {
      this._value1Stack = this._value1Stack.pop();
      this._value2Stack = this._value2Stack.pop();
    }
    super.leave(node);
  }

  getValue1(): mixed {
    return this._value1Stack.peek();
  }

  getValue2(): mixed {
    return this._value2Stack.peek();
  }

  shouldBreak(): boolean {
    return this._shouldBreak;
  }
}

export function selectedDataHaveChanged({
  query,
  schema,
  queryHash,
  reducer1,
  reducer2,
}: {
  query: string,
  schema: GraphQLSchema,
  queryHash: string | number,
  reducer1: DataReducerRecord,
  reducer2: DataReducerRecord,
}): boolean {
  const typeInfo = new TypeInfoWithValuesComparator({
    schema,
    reducer1,
    reducer2,
    queryHash1: queryHash,
    queryHash2: queryHash,
  });
  let hasChanged = false;
  visit(
    parse(query),
    visitWithTypeInfo(typeInfo, {
      [Kind.FIELD]: {
        enter(node) {
          hasChanged = typeInfo.shouldBreak();
        },
      },
    }),
  );
  return hasChanged;
}
