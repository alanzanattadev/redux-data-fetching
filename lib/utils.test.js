import {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLString,
  GraphQLList,
  GraphQLInt,
  visit,
  visitWithTypeInfo,
  parse,
  TypeInfo,
  getNamedType,
} from "graphql";
import { Map, List, Record } from "immutable";
import { selectedDataHaveChanged } from "./utils";
import { DataReducerRecord, ResultsRecord, QueryRecord } from "./reducer";

const Event = new GraphQLObjectType({
  name: "Event",
  fields: () => ({
    timestamp: { type: GraphQLInt },
    triggered_by: { type: Boss },
  }),
});

const User = new GraphQLObjectType({
  name: "User",
  fields: () => ({
    id: { type: GraphQLString },
    name: { type: GraphQLString },
    description: { type: GraphQLString },
    friends: { type: new GraphQLList(User) },
    boss: { type: Boss },
    events: { type: new GraphQLList(Event) },
  }),
});

const Boss = new GraphQLObjectType({
  name: "Boss",
  fields: () => ({
    id: { type: GraphQLString },
    company: { type: GraphQLString },
  }),
});

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: "RootQuery",
    fields: {
      user: {
        type: User,
        resolve: () => {},
        args: { id: { type: GraphQLString } },
      },
      users: {
        type: new GraphQLList(User),
        resolve: (p, args, context) =>
          context.db && context.queryHash ? [] : null,
      },
      boss: {
        type: Boss,
        resolve: () => boss1,
      },
    },
  }),
});

const UserRecord = Record({
  id: null,
  name: null,
  description: null,
  friends: null,
  boss: null,
  events: null,
});

const BossRecord = Record({
  id: null,
  company: null,
});

const EventRecord = Record({
  timestamp: 0,
  triggered_by: null,
});

const user1 = new UserRecord({
  id: 1,
  boss: 2,
  friends: [1],
  events: [
    new EventRecord({ triggered_by: 2 }),
    new EventRecord(),
    new EventRecord(),
  ],
});

const boss1 = new BossRecord({
  id: 2,
  company: "Microsoft",
});

const query = new QueryRecord({
  results: new ResultsRecord({
    byQuery: Map({
      users: [1],
      boss: 2,
    }),
  }),
});

const hash = "hash";
const queries = Map({ [hash]: query });

const reducer1 = new DataReducerRecord({
  entities: Map({
    User: Map({
      "1": user1,
    }),
    Boss: Map({
      "2": boss1,
    }),
  }),
  queries,
});

const reducer2 = new DataReducerRecord({
  entities: Map({
    User: Map({
      "1": user1,
    }),
    Boss: Map({
      "2": boss1.set("company", "Google"),
    }),
  }),
  queries,
});

describe("Utils", () => {
  describe("selectedDataHaveChanged", () => {
    it("can detect change correctly", () => {
      expect(
        selectedDataHaveChanged({
          query: `{ users { id, friends { id, name, friends { id, events { timestamp, triggered_by { company } } } }, boss { company } } }`,
          schema,
          reducer1,
          reducer2,
          queryHash: hash,
        }),
      ).toEqual(true);

      expect(
        selectedDataHaveChanged({
          query: `{ users { id, friends { id, name, friends { id, events { timestamp, triggered_by { company } } } } } }`,
          schema,
          reducer1,
          reducer2,
          queryHash: hash,
        }),
      ).toEqual(true);

      expect(
        selectedDataHaveChanged({
          query: `{ users { id, friends { id, name, friends { id, events { timestamp } } } } }`,
          schema,
          reducer1,
          reducer2,
          queryHash: hash,
        }),
      ).toEqual(false);

      expect(
        selectedDataHaveChanged({
          query: `{ boss { company } }`,
          schema,
          reducer1,
          reducer2,
          queryHash: hash,
        }),
      ).toEqual(true);
    });
  });
});
