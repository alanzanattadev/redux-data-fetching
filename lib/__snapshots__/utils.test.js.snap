// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils hashMutationQuery creates a hash 1`] = `653432890`;
